### PR TITLE
Replace to tcell from termbox

### PR DESCRIPTION
**Pull request by tjmtmmnk**
Related Issue: #15 

Thanks great library!
This PR is intended to replace completely from termbox to tcell.

I think this migration has two merits.

1. It is stated that termbox will not be maintained anymore ([ref](https://github.com/nsf/termbox-go)). So, migrate to maintained tcell, make this product more stable.

2. tcell has a lot of useful features. For example, #29 can solve easily by using code used in tview etc ([ref](https://github.com/rivo/tview/blob/master/ansi.go)).

---

The problem reported in #17, can solve by using workaround([ref](https://github.com/gdamore/tcell/issues/194#issuecomment-478209769)).

I could confirm [the code](https://gist.github.com/akerl-unpriv/d94d02cb21a333776d4ebaf2a24f8849)(reproduce the problem) work well in local.

### CODE REVIEW

**Comment by ktr0731** [comment]
Hi, @tjmtmmnk. Thanks a lot for your great work!
I'll review this Pull Request later, so please take a little...

File: fuzzyfinder.go [hunk 1]
```diff
@@ -553,7 +618,6 @@ func (f *finder) find(slice interface{}, itemFunc func(i int) string, opts []Opt
 	if err := f.initFinder(items, matched, opt); err != nil {
 		return nil, errors.Wrap(err, "failed to initialize the fuzzy finder")
 	}
-	defer f.term.close()
```
Comment (ktr0731) [inline fuzzyfinder.go:556]: Should we call `f.term.Screen().Fini()` at the end?

File: fuzzyfinder.go [hunk 2]
```diff
@@ -604,6 +672,20 @@ func (f *finder) find(slice interface{}, itemFunc func(i int) string, opts []Opt
 	}
 }
 
+// This method avoid tcell bug https://github.com/gdamore/tcell/issues/194
+// Aditional EOL event is sent to ensure, consequent events, are correctly handled
+func sendExtraEventFix() {
+	kb, err := keybd_event.NewKeyBonding()
+	if err != nil {
+		panic(err)
```
Comment (ktr0731) [inline fuzzyfinder.go:680]: Do not use panic, return the error instead.

File: fuzzyfinder.go [hunk 3]
```diff
@@ -604,6 +672,20 @@ func (f *finder) find(slice interface{}, itemFunc func(i int) string, opts []Opt
 	}
 }
 
+// This method avoid tcell bug https://github.com/gdamore/tcell/issues/194
+// Aditional EOL event is sent to ensure, consequent events, are correctly handled
+func sendExtraEventFix() {
```
Comment (ktr0731) [inline fuzzyfinder.go:677]: Great hack 👍

File: fuzzyfinder.go [hunk 4]
```diff
@@ -604,6 +672,20 @@ func (f *finder) find(slice interface{}, itemFunc func(i int) string, opts []Opt
 	}
 }
 
+// This method avoid tcell bug https://github.com/gdamore/tcell/issues/194
+// Aditional EOL event is sent to ensure, consequent events, are correctly handled
```
Comment (ktr0731) [inline fuzzyfinder.go:676]:
```suggestion
// Additional EOL event is sent to ensure, consequent events, are correctly handled.
```

File: fuzzing_test.go [hunk 5]
```diff
@@ -1,5 +1,3 @@
-// +build fuzz
```
Comment (ktr0731) [inline fuzzing_test.go:1]: I had deliberately specified this build constraint. Could you revert this line?

File: fuzzing_test.go [hunk 6]
```diff
@@ -87,25 +84,32 @@ func TestFuzz(t *testing.T) {
...
-		n := rand.Intn(*numEvents) + 10
-		events := make([]termbox.Event, n)
+	min := func(a, b int) int {
+		if a < b {
+			return a
+		}
+		return b
+	}
+
+	// number of events in tcell.SimulationScreen is limited 10
```
Comment (ktr0731) [inline fuzzing_test.go:94]: `*numCases` doesn't seem to be related to the limitation of the number of events, is it right?

**Review by ktr0731 — COMMENTED** [review]
Thanks for your contribution! I commented on some points.

File: fuzzing_test.go (hunk 5 above)
Comment (tjmtmmnk) [reply fuzzing_test.go:1]: sorry, mis-comitted...

File: fuzzing_test.go (hunk 6 above)
Comment (tjmtmmnk) [reply fuzzing_test.go:94]: tcell simulation screen limit 10 events https://github.com/gdamore/tcell/blob/master/simulation.go#L112

File: mock.go [hunk 7]
```diff
@@ -4,231 +4,166 @@ import (
...
+	screen   tcell.SimulationScreen
 	resultMu sync.RWMutex
 	result   string
-
-	sleepDuration time.Duration
-}
-
-// SetSize changes the pseudo-size of the window.
-// Note that SetSize resets added cells.
-func (m *TerminalMock) SetSize(w, h int) {
```
[the remark this answers is not in the document]
Comment (tjmtmmnk) [reply mock.go:38]: ok. I will define `TerminalMockV2`.

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -30,7 +30,7 @@ jobs:
       run: go build
 
     - name: Test
-      run: go test -v -coverpkg ./... -covermode atomic -coverprofile coverage.txt -tags fuzz -numCases 3000 -numEvents 300 ./...
```
[the remark this answers is not in the document]
Comment (tjmtmmnk) [reply .github/workflows/main.yml:33]: This change related to https://github.com/ktr0731/go-fuzzyfinder/pull/32#discussion_r535270597

File: fuzzyfinder.go (hunk 1 above)
Comment (tjmtmmnk) [reply fuzzyfinder.go:556]: I call `Fini()` [here](https://github.com/ktr0731/go-fuzzyfinder/pull/32/files?file-filters%5B%5D=.go#diff-902eb73b67dfd52495aa77c7330f4734c5617b00a9faaebb99c97ec583a927c3R706) and [here](https://github.com/ktr0731/go-fuzzyfinder/pull/32/files?file-filters%5B%5D=.go#diff-902eb73b67dfd52495aa77c7330f4734c5617b00a9faaebb99c97ec583a927c3R726)

File: mock.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tjmtmmnk) [reply mock.go:38]: To me `TerminalMock` seemed to be used only for testing purposes.
I didn't understand why backwards compatibility was needed so please let me know...

[the remark this answers is not in the document]
Comment (ktr0731) [reply mock.go:38]: Yes, it is used only for testing. However, according to the [import compatibility rule](https://blog.golang.org/v2-go-modules) of Go modules, public interfaces should be compatible, regardless of whether they are for testing or not.

Please let me know if there seems to be no way to keep backward compatibility; I will consider making a v2 release...

File: fuzzing_test.go (hunk 6 above)
Comment (tjmtmmnk) [reply fuzzing_test.go:96]: Oh, that's right🙏

File: fuzzyfinder.go (hunk 1 above)
Comment (tjmtmmnk) [reply fuzzyfinder.go:556]: I was misunderstanding, as you say, I can call `Fini()` in `find`

File: mock.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tjmtmmnk) [reply mock.go:38]: Thanks for introducing useful article!
I tried to make the exposed interface backwards compatible.

**Comment by tjmtmmnk** [comment]
- tried to make the public interface backwards compatible by creating v2
- embedded `struct` implementing `Screen` or `SimulationScreen` interface
- call `Fini` in `find`

I fixed 3 points.
Please re-review when you have a time🙏

File: helper_test.go [hunk 9]
```diff
@@ -11,3 +11,11 @@ func NewWithMockedTerminal() (*finder, *TerminalMock) {
 	m.SetSize(w, h)
 	return f, m
 }
+
+func NewWithMockedTerminalV2() (*finder, *TerminalMock) {
```
Comment (ktr0731) [inline helper_test.go:15]: There is no need to define a v2 function because `helper_test.go` is only for testing and it is referenced only from `fuzzyfinder_test.go`. So you can rename it to `NewWithMockedTerminal` and remove the old one.

```suggestion
func NewWithMockedTerminal() (*finder, *TerminalMock) {
```

File: mock.go [hunk 10]
```diff
@@ -78,26 +135,146 @@ func (t *TerminalMock) GetResult() string {
...
+	defer m.resultMu.Unlock()
+	m.result = s
+
+	return nil
+}
+
+func (m *TerminalMock) close() {}
+
+// UseMockedTerminal switches the terminal, which is used from
+// this package to a mocked one.
```
Comment (ktr0731) [inline mock.go]: Could you add a comment for deprecation? (ref. https://github.com/golang/go/wiki/Deprecated)

```suggestion
// this package to a mocked one.
//
// Deprecated: Use UseMockedTerminalV2.
```

File: mock.go [hunk 11]
```diff
@@ -78,26 +135,146 @@ func (t *TerminalMock) GetResult() string {
...
+func UseMockedTerminal() *TerminalMock {
+	return defaultFinder.UseMockedTerminal()
+}
+
```
Comment (ktr0731) [inline mock.go:266]:
```suggestion

// UseMockedTerminalV2 switches the terminal, which is used from
// this package to a mocked one.
```

File: mock.go [hunk 12]
```diff
@@ -109,8 +286,50 @@ func (f *finder) UseMockedTerminal() *TerminalMock {
...
+		} else {
+			fmt.Fprintf(&buf, "48;5;%d", color)
+		}
+	}
+	buf.WriteString("m")
+
+	return buf.String()
+}
+
+// parseAttrV2 parses color and attribute for test
```
Comment (ktr0731) [inline mock.go:331]: [super nits] 
```suggestion
// parseAttrV2 parses color and attribute for testing.
```

File: mock.go [hunk 13]
```diff
@@ -45,12 +49,30 @@ func (m *TerminalMock) SetSize(w, h int) {
 	m.cells = make([]*cell, w*h)
 }
 
+func (m *TerminalMock) SetSizeV2(w, h int) {
```
Comment (ktr0731) [inline mock.go:52]: Could you switch the implementation of `SetSize` based on a `bool` field such as `m.v2` instead of defining `SetSizeV2`? 

For example:

``` go
// SetSize changes the pseudo-size of the window.
// Note that SetSize resets added cells.
func (m *TerminalMock) SetSize(w, h int) {
	if m.v2 {
		m.simScreen.SetSize(w, h)
		return
	}

	m.sizeMu.Lock()
	defer m.sizeMu.Unlock()
	m.cellsMu.Lock()
	defer m.cellsMu.Unlock()
	m.width = w
	m.height = h
	m.cells = make([]*cell, w*h)
}
```

File: mock.go [hunk 14]
```diff
@@ -45,12 +49,30 @@ func (m *TerminalMock) SetSize(w, h int) {
...
+	m.simScreen.SetSize(w, h)
+}
+
 // SetEvents sets all events, which are fetched by pollEvent.
 // A user of this must set the EscKey event at the end.
-func (m *TerminalMock) SetEvents(e ...termbox.Event) {
+func (m *TerminalMock) SetEvents(events ...termbox.Event) {
```
Comment (ktr0731) [inline mock.go:58]: 👍

Comment (ktr0731) [inline mock.go:58]:
```suggestion
// Deprecated: Use SetEventsV2.
func (m *TerminalMock) SetEvents(events ...termbox.Event) {
```

File: mock.go [hunk 15]
```diff
@@ -45,12 +49,30 @@ func (m *TerminalMock) SetSize(w, h int) {
...
 	m.eventsMu.Lock()
 	defer m.eventsMu.Unlock()
-	m.events = e
+	m.events = events
+}
+
```
Comment (ktr0731) [inline mock.go:63]:
```suggestion

// SetEventsV2 sets all events, which are fetched by pollEvent.
// A user of this must set the EscKey event at the end.
```

File: mock.go [hunk 16]
```diff
@@ -61,6 +83,63 @@ func (m *TerminalMock) GetResult() string {
 	return m.result
 }
 
+func (m *TerminalMock) GetResultV2() string {
```
Comment (ktr0731) [inline mock.go:86]: Same as the review comment on `SetSizeV2`.

**Review by ktr0731 — COMMENTED** [review]
@tjmtmmnk
I'm really sorry I'm always late...

It's almost perfect, but I'd like you to fix a few minor things before we merge it 🙏

**Comment by tjmtmmnk** [comment]
Sorry for the delay, and thank you for review!
I fixed problems you indicated, please check when you have a time🙏

**Review by ktr0731 — APPROVED** [review]
LGTM! 
Thanks for your contribution. I'm sure that this change will be very helpful for `go-fuzzyfinder`'s future.
Have a happy new year holiday 🍺

**Comment by tjmtmmnk** [comment]
Thank you!!
